Segfault in master during `lookupByID()`

I installed v1.7, then upgraded to v1.8, then upgraded to master (`docker.io/cilium/cilium:latest`) as outlined in my comment here:
https://github.com/cilium/cilium/pull/13259#issuecomment-705894240

The connectivity check was running in the background, including the `pod-to-external-fqdn-allow-google-cnp` pod with DNS + FQDN policy applied.

When upgrading to master, on 1/2 nodes Cilium failed out the first couple of times before successfully running. In the previous logs, there was this error at the end:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1b5bae1]

goroutine 314 [running]:
github.com/cilium/cilium/pkg/identity/cache.(*localIdentityCache).lookupByID(0x0, 0xc0000023d9, 0x0)
        /go/src/github.com/cilium/cilium/pkg/identity/cache/local.go:168 +0x41
github.com/cilium/cilium/pkg/identity/cache.(*CachingIdentityAllocator).LookupIdentityByID(0xc00096a2a0, 0x298ec00, 0xc00017f640, 0x23d9, 0x25ea694)
        /go/src/github.com/cilium/cilium/pkg/identity/cache/cache.go:245 +0x86
github.com/cilium/cilium/daemon/cmd.(*Daemon).notifyOnDNSMsg.func3(0xc000f001c0)
        /go/src/github.com/cilium/cilium/daemon/cmd/fqdn.go:471 +0x5ef
github.com/cilium/cilium/pkg/proxy/logger.NewLogRecord(0x2947340, 0x3fddbc0, 0x29b9cc0, 0xc000978900, 0x25cdc35, 0x7, 0xffffff00, 0xc000f79ae0, 0x5, 0x5, ...)
        /go/src/github.com/cilium/cilium/pkg/proxy/logger/logger.go:110 +0x28e
github.com/cilium/cilium/daemon/cmd.(*Daemon).notifyOnDNSMsg(0xc00000a1e0, 0xbfd8083315353350, 0xa66aed71, 0x3a817a0, 0xc000978900, 0xc000501690, 0xf, 0xc0005016f0, 0xc, 0xc0003ec750, ...)
        /go/src/github.com/cilium/cilium/daemon/cmd/fqdn.go:439 +0xca5
github.com/cilium/cilium/pkg/fqdn/dnsproxy.(*DNSProxy).ServeDNS(0xc0003ceb40, 0x29ae500, 0xc000407200, 0xc0003ec750)
        /go/src/github.com/cilium/cilium/pkg/fqdn/dnsproxy/proxy.go:570 +0x171a
github.com/miekg/dns.(*Server).serveDNS(0xc00014c5a0, 0xc000d1b800, 0x3a, 0x200, 0xc000407200)
        /go/src/github.com/cilium/cilium/vendor/github.com/miekg/dns/server.go:597 +0x211
github.com/miekg/dns.(*Server).serveUDPPacket(0xc00014c5a0, 0xc0003f20b0, 0xc000d1b800, 0x3a, 0x200, 0xc000bc6120, 0xc000599a50)
        /go/src/github.com/cilium/cilium/vendor/github.com/miekg/dns/server.go:552 +0xb2
github.com/miekg/dns.(*Server).serveUDP.func2(0xc00014c5a0, 0xc0003f20b0, 0xc000d1b800, 0x3a, 0x200, 0xc000bc6120, 0xc000599a50)
        /go/src/github.com/cilium/cilium/vendor/github.com/miekg/dns/server.go:478 +0x67
created by github.com/miekg/dns.(*Server).serveUDP
        /go/src/github.com/cilium/cilium/vendor/github.com/miekg/dns/server.go:477 +0x24f
```

After two restarts, the problem seemed to resolve itself.

--- a/pkg/identity/cache/cache.go
+++ b/pkg/identity/cache/cache.go
@@ -199,19 +199,46 @@ func (w *identityWatcher) stop() {
 	close(w.stopChan)
 }
 
+// isLocalIdentityAllocatorInitialized returns true if m.localIdentities is not nil.
+func (m *CachingIdentityAllocator) isLocalIdentityAllocatorInitialized() bool {
+	select {
+	case <-m.localIdentityAllocatorInitialized:
+		return m.localIdentities != nil
+	default:
+		return false
+	}
+}
+
+// isGlobalIdentityAllocatorInitialized returns true if m.IdentityAllocator is not nil.
+// Note: This does not mean that the identities have been synchronized,
+// see WaitForInitialGlobalIdentities to wait for a fully populated cache.
+func (m *CachingIdentityAllocator) isGlobalIdentityAllocatorInitialized() bool {
+	select {
+	case <-m.globalIdentityAllocatorInitialized:
+		return m.IdentityAllocator != nil
+	default:
+		return false
+	}
+}
+
 // LookupIdentity looks up the identity by its labels but does not create it.
 // This function will first search through the local cache, then the caches for
-// remote kvstores and finally fall back to the main kvstore
+// remote kvstores and finally fall back to the main kvstore.
+// May return nil for lookups if the allocator has not yet been synchronized.
 func (m *CachingIdentityAllocator) LookupIdentity(ctx context.Context, lbls labels.Labels) *identity.Identity {
 	if reservedIdentity := identity.LookupReservedIdentityByLabels(lbls); reservedIdentity != nil {
 		return reservedIdentity
+	}
+
+	if !m.isLocalIdentityAllocatorInitialized() {
+		return nil
 	}
 
 	if identity := m.localIdentities.lookup(lbls); identity != nil {
 		return identity
 	}
 
-	if !identity.RequiresGlobalIdentity(lbls) || m.IdentityAllocator == nil {
+	if !identity.RequiresGlobalIdentity(lbls) || !m.isGlobalIdentityAllocatorInitialized() {
 		return nil
 	}
 
@@ -233,6 +260,7 @@ var unknownIdentity = identity.NewIdentity(identity.IdentityUnknown, labels.Labe
 // LookupIdentityByID returns the identity by ID. This function will first
 // search through the local cache, then the caches for remote kvstores and
 // finally fall back to the main kvstore
+// May return nil for lookups if the allocator has not yet been synchronized.
 func (m *CachingIdentityAllocator) LookupIdentityByID(ctx context.Context, id identity.NumericIdentity) *identity.Identity {
 	if id == identity.IdentityUnknown {
 		return unknownIdentity
@@ -242,11 +270,15 @@ func (m *CachingIdentityAllocator) LookupIdentityByID(ctx context.Context, id id
 		return identity
 	}
 
+	if !m.isLocalIdentityAllocatorInitialized() {
+		return nil
+	}
+
 	if identity := m.localIdentities.lookupByID(id); identity != nil {
 		return identity
 	}
 
-	if id.HasLocalScope() || m.IdentityAllocator == nil {
+	if !m.isGlobalIdentityAllocatorInitialized() || id.HasLocalScope() {
 		return nil
 	}
 

--- a/pkg/identity/cache/cache.go
+++ b/pkg/identity/cache/cache.go
@@ -51,8 +51,7 @@ func (m *CachingIdentityAllocator) GetIdentityCache() IdentityCache {
 	log.Debug("getting identity cache for identity allocator manager")
 	cache := IdentityCache{}
 
-	if m.IdentityAllocator != nil {
-
+	if m.isGlobalIdentityAllocatorInitialized() {
 		m.IdentityAllocator.ForeachCache(func(id idpool.ID, val allocator.AllocatorKey) {
 			if val != nil {
 				if gi, ok := val.(GlobalIdentity); ok {
@@ -69,7 +68,7 @@ func (m *CachingIdentityAllocator) GetIdentityCache() IdentityCache {
 		cache[key] = identity.Labels.LabelArray()
 	}
 
-	if m.localIdentities != nil {
+	if m.isLocalIdentityAllocatorInitialized() {
 		for _, identity := range m.localIdentities.GetIdentities() {
 			cache[identity.ID] = identity.Labels.LabelArray()
 		}
@@ -82,22 +81,25 @@ func (m *CachingIdentityAllocator) GetIdentityCache() IdentityCache {
 func (m *CachingIdentityAllocator) GetIdentities() IdentitiesModel {
 	identities := IdentitiesModel{}
 
-	m.IdentityAllocator.ForeachCache(func(id idpool.ID, val allocator.AllocatorKey) {
-		if gi, ok := val.(GlobalIdentity); ok {
-			identity := identity.NewIdentityFromLabelArray(identity.NumericIdentity(id), gi.LabelArray)
-			identities = append(identities, identitymodel.CreateModel(identity))
-		}
+	if m.isGlobalIdentityAllocatorInitialized() {
+		m.IdentityAllocator.ForeachCache(func(id idpool.ID, val allocator.AllocatorKey) {
+			if gi, ok := val.(GlobalIdentity); ok {
+				identity := identity.NewIdentityFromLabelArray(identity.NumericIdentity(id), gi.LabelArray)
+				identities = append(identities, identitymodel.CreateModel(identity))
+			}
 
-	})
+		})
+	}
 	// append user reserved identities
 	for _, v := range identity.ReservedIdentityCache {
 		identities = append(identities, identitymodel.CreateModel(v))
 	}
 
-	for _, v := range m.localIdentities.GetIdentities() {
-		identities = append(identities, identitymodel.CreateModel(v))
+	if m.isLocalIdentityAllocatorInitialized() {
+		for _, v := range m.localIdentities.GetIdentities() {
+			identities = append(identities, identitymodel.CreateModel(v))
+		}
 	}
-
 	return identities
 }
 

